src/parthenon_arrays.hpp

```diff
@@ -0,0 +1,315 @@
+//========================================================================================
+// Parthenon++ astrophysical MHD code
+// Copyright(C) 2014 James M. Stone <jmstone@princeton.edu> and other code contributors
+// Licensed under the 3-clause BSD License, see LICENSE file for details
+//========================================================================================
+// (C) (or copyright) 2020. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef PARTHENON_ARRAYS_HPP_
+#define PARTHENON_ARRAYS_HPP_
+
+// C++ headers
+#include <cassert>
+#include <cstddef> // size_t
+#include <string>
+#include <type_traits>
+#include <utility> // make_pair
+#include <vector>
+
+// Kokkos Headers
+#include <Kokkos_Core.hpp>
+
+// Parthenon++ headers
+#include "kokkos_abstraction.hpp"
+
+// Macro for automatically creating a useful name
+#define PARARRAY_TEMP\
+  "ParArrayNDGeneric:"+std::string(__FILE__)+":"+std::to_string(__LINE__)
+
+namespace parthenon {
+
+// API designed with Container = Kokkos::View<T******> in mind
+template <typename Container>
+class ParArrayNDGeneric {
+ public:
+  using index_pair_t = std::pair<size_t,size_t>;
+
+  KOKKOS_INLINE_FUNCTION
+  ParArrayNDGeneric() = default;
+  ParArrayNDGeneric(const std::string& label,
+                      int nx6, int nx5, int nx4, int nx3, int nx2, int nx1)
```

Comment:
Interesting. I assumed `Kokkos` would check. But I just tried, and it doesn't. `Kokkos` silently casts the `int` to a `size_t` and so you get integer overflow if you pass in a negative integer. I'll add some asserts.

Reply:
Confirmed that the same problem occurs for `operator()`. Passing in a negative integer causes a segfault. I don't feel bad about adding asserts, since we build with `-DNDEBUG` for production runs anyway.

Reply:
Bounds checking is enablable for Kokkos views but you must turn it on.

Reply:
Ah. How do I do that? If that's the case, I should get rid of all those asserts I just added.

Reply:
If you set KOKKOS_DEBUG=yes. There is a camel command but I don’t have it in front of me. There is another more explicit command as well but note that it will take a heavy performance hit.

Reply:
As long as bounds get checked in Kokkos or Parthenon when we compile with Debug, I'm happy. We might need to tweak CMake to make sure we get Kokkos in debug mode if we compile debug